tools/cell_census_builder/validate.py

```diff
@@ -361,73 +364,81 @@ def validate_X_layers(
     Raises on error.  Returns True on success.
     """
     logging.debug("validate_X_layers")
-    with soma.Collection.open(soma_path, context=SOMA_TileDB_Context()) as census:
-        census_data = census[CENSUS_DATA_NAME]
-
-        for eb in experiment_builders:
-            se = census_data[eb.name]
-            assert soma.Collection.exists(se.ms["RNA"].X.uri)
-
-            census_obs_df = se.obs.read(column_names=["soma_joinid"]).concat().to_pandas()
-            n_obs = len(census_obs_df)
-            logging.info(f"uri = {se.obs.uri}, eb.n_obs = {eb.n_obs}; n_obs = {n_obs}")
-            # TODO: Only works when run as builder, not standalone validator, since ExpBuilder is not fully initialized
-            #  in the latter case
+    for eb in reopen_experiment_builders(experiment_builders, 'r'):
```

Comment:
the need to reopen concerns me.  It implies that there is insufficient isolation between the census "build" code and the "validate" code.

I would have expected that the validation pass would work off a clean, newly opened, for read, set of experiment handles.  And therefore would have no possible SOMA shared state, and therefore no need to re-open.

Anything where there is shared SOMA state, eg., open handles, seems like it is doing to potentially be a suspect validation.

Note: I think it is _fine_ to utilize the parts of the ExperimentBuilder that is declarative input to the builder (ie., the name, filter spec, gene_feature stuff, etc).  But the reopen suggests much more is shared, which makes me not trust the validation.

Reply:
Thinking about this some more -- what has happened (I started it, you added to it) is that the ExperimentBuilder has grown to be two things:
* declaration of the Experiment "definition" -- filters, names, gene_length, etc.
* accumulated state used to "build" the experiment.  eg., var accumulator, n_obs, etc.

The build phase needs both.  The validator should only use the first item, or it isn't really validating in an isolated manner.

Stuff like use of `reopen...` in the validator is a sign that accumulated build state may be creeping into the validator.  Not ideal IMHO.

Not sure it should hold up this PR, but I think a better split here would be a good idea.  Concretely, we could solve this by splitting the ExperimentBuilder object into:
* ExperimentDefinition - containing name, filter, gene_length stuff, etc. 
* ExperimentBuilder - subclass of ExperimentDefinition, adding the build-time accumulators for of obs/var, open soma.Experiment handles, etc.

the validator would be given only the first (definition).

Reply:
understood and agreed; this the root problem of https://github.com/chanzuckerberg/cell-census/issues/179.  linked back to this thread from the issue, so we can address all this properly on that issue.

Reply:
The other place we lack sufficient isolation is the re-use of the SOMA/TileDB context.  The validator should have its own, freshly created context.

Reply:
filed as #215